### PR TITLE
Feature/507 flags config refactor

### PR DESCRIPTION
**Pull request by potehinre**
refactoring of config and flags

### CODE REVIEW

File: packages/daemons/block_generator.go [hunk 1]
```diff
@@ -132,7 +132,7 @@ func BlockGenerator(ctx context.Context, d *daemon) error {
 	header := &utils.BlockData{
 		BlockID:      prevBlock.BlockID + 1,
 		Time:         time.Now().Unix(),
-		EcosystemID:  conf.Config.EcosystemID,
+		EcosystemID:  0,
```
Comment (ghost) [inline packages/daemons/block_generator.go:141]: what is the purpose of this field if it is always zero?

File: tools/generate_first_block/main.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		log.WithFields(log.Fields{"value": *publicKey, "error": err}).Fatal("converting public key from hex")
+	}
+	bNodePublicKey, err := hex.DecodeString(*nodePublicKey)
+	if err != nil {
+		log.WithFields(log.Fields{"value": *nodePublicKey, "error": err}).Fatal("converting node public key from hex")
+	}
+	_, err = converter.BinMarshal(&tx,
+		&consts.FirstBlock{
+			TxHeader: consts.TxHeader{
+				// TODO: move types to enum
```
Comment (ghost) [inline tools/generate_first_block/main.go:48]: We decided not to write todo's but rewrite code. Will it take a lot of time to create this enum?

**Review by ghost — APPROVED** [review]

File: packages/daylight/start.go [hunk 3]
```diff
@@ -121,11 +120,28 @@ func savePid() error {
...
 func delPidFile() {
-	os.Remove(conf.GetPidFile())
+	os.Remove(conf.Config.GetPidPath())
+}
+
+func DelLockFile() error {
+	return os.Remove(conf.Config.LockFilePath)
+}
+
+func IsLockFileExists() bool {
```
Comment (potehinre) [inline packages/daylight/start.go]: this is not correct at all, unix-style file locking should use fcntl.Flock syscall.
For example see https://github.com/tgulacsi/go-locking/blob/master/locking.go

File: packages/conf/conf.go [hunk 4]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
 
-// NoConfig config file does not exist
-func NoConfig() bool {
-	_, err := os.Stat(GetConfigPath())
-	return os.IsNotExist(err)
+func FillRuntimePaths() error {
+	if Config.DataDir == "" {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return errors.Wrapf(err, "getting current wd")
```
Comment (potehinre) [inline packages/conf/conf.go:164]: logging?

File: packages/conf/conf.go [hunk 5]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
+		}
+
+		Config.DataDir = filepath.Join(cwd, "genesis-data")
```
Comment (potehinre) [inline packages/conf/conf.go]: "genesis-data" should be const

File: packages/conf/conf.go [hunk 6]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
 
 // SaveConfig save global parameters to configFile
-func SaveConfig() error {
-	cf, err := os.Create(GetConfigPath())
+func SaveConfig(path string) error {
+	dir := filepath.Dir(path)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err := os.Mkdir(dir, 0775)
+		if err != nil {
+			return errors.Wrapf(err, "creating dir %s", dir)
```
Comment (potehinre) [inline packages/conf/conf.go:142]: logging?

File: packages/conf/conf.go [hunk 7]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
+
+	viper.SetConfigFile(path)
+	err = viper.ReadInConfig()
+	if err != nil {
+		return errors.Wrapf(err, "reading config")
+	}
+
+	err = viper.Unmarshal(&Config)
+	if err != nil {
+		return errors.Wrapf(err, "marshalling config to global struct variable")
```
Comment (potehinre) [inline packages/conf/conf.go]: logging?

File: packages/conf/conf.go [hunk 8]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
+
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return errors.Errorf("Unable to load config file %s", path)
+	}
+
+	viper.SetConfigFile(path)
+	err = viper.ReadInConfig()
+	if err != nil {
+		return errors.Wrapf(err, "reading config")
```
Comment (potehinre) [inline packages/conf/conf.go]: logging?

File: cmd/root.go [hunk 9]
```diff
@@ -1,30 +1,67 @@
...
+	}
+}
+
+func defautConfigPath() string {
+	p, err := os.Getwd()
+	if err != nil {
+		log.Fatal("getting cur wd")
+	}
+
+	return filepath.Join(p, "genesis-data", "config.toml")
```
Comment (potehinre) [inline cmd/root.go]: "genesis-data", "config.toml" should be consts

File: cmd/generateFirstBlock.go [hunk 10]
```diff
@@ -7,55 +7,60 @@ import (
...
+		decodeKeyFile := func(kName string) []byte {
+			filepath := filepath.Join(conf.Config.KeysDir, kName)
+			data, err := ioutil.ReadFile(filepath)
+			if err != nil {
+				log.WithError(err).Fatalf("Reading %s key data from %s", kName, filepath)
+			}
+
+			decodedKey, err := hex.DecodeString(string(data))
+			if err != nil {
+				log.WithError(err).Fatal("converting %s from hex", kName)
```
Comment (potehinre) [inline cmd/generateFirstBlock.go]: log.WithFields(log.Fields{"type": consts.ConversionError, "value": kName, "error": err}).Error("converting value from hex")

File: cmd/generateFirstBlock.go [hunk 11]
```diff
@@ -7,55 +7,60 @@ import (
...
-			log.WithFields(log.Fields{"value": publicKey, "error": err}).Fatal("converting public key from hex")
-		}
-		bNodePublicKey, err := hex.DecodeString(nodePublicKey)
-		if err != nil {
-			log.WithFields(log.Fields{"value": nodePublicKey, "error": err}).Fatal("converting node public key from hex")
+		decodeKeyFile := func(kName string) []byte {
+			filepath := filepath.Join(conf.Config.KeysDir, kName)
+			data, err := ioutil.ReadFile(filepath)
+			if err != nil {
+				log.WithError(err).Fatalf("Reading %s key data from %s", kName, filepath)
```
Comment (potehinre) [inline cmd/generateFirstBlock.go]: Formatted logging is prohibited, use WithFields

File: cmd/config.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+			log.Fatalf("Filling config: %s", err.Error())
+		}
+
+		if configPath == "" {
+			configPath = filepath.Join(conf.Config.DataDir, "config.toml")
+		}
+
+		err = conf.SaveConfig(configPath)
+		if err != nil {
+			log.Fatalf("Saving config error: %s", err.Error())
```
Comment (potehinre) [inline cmd/config.go]: WithError(err)

File: cmd/config.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+var configCmd = &cobra.Command{
+	Use:   "config",
+	Short: "Initial config generation",
+	Run: func(cmd *cobra.Command, args []string) {
+		// Error omitted because we have default flag value
+		configPath, _ := cmd.Flags().GetString("path")
+
+		err := conf.FillRuntimePaths()
+		if err != nil {
+			log.Fatalf("Filling config: %s", err.Error())
```
Comment (potehinre) [inline cmd/config.go]: WithError

File: packages/conf/conf.go [hunk 14]
```diff
@@ -48,78 +55,138 @@ type LogConfig struct {
...
 	if err != nil {
 		log.WithFields(log.Fields{"type": consts.IOError, "error": err}).Error("Create config file failed")
 		return err
 	}
 	defer cf.Close()
-	return toml.NewEncoder(cf).Encode(Config)
+
+	err = toml.NewEncoder(cf).Encode(Config)
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/conf/conf.go:155]: logging?

File: packages/daylight/start.go [hunk 15]
```diff
@@ -225,6 +219,13 @@ func Start() {
 	}
 	defer delPidFile()
 
+	// create lock file
+	if err := CreateLockFile(); err != nil {
+		log.Errorf("can't create lock: %s", err)
```
Comment (potehinre) [inline packages/daylight/start.go]: WithError(err)

File: packages/conf/conf.go (hunk 4 above)
Comment (innwkd) [reply packages/conf/conf.go:164]: This error is logged in the parent function

File: packages/conf/conf.go (hunk 14 above)
Comment (innwkd) [reply packages/conf/conf.go:155]: This error is logged in the parent function

**Pull request merged**